cogeo_mosaic/backend/dynamo.py

```diff
@@ -0,0 +1,88 @@
+import functools
+import itertools
+import os
+from typing import Dict, Tuple
+
+import mercantile
+
+from cogeo_mosaic.backend.base import BaseBackend
+
+
+class DynamoDBBackend(BaseBackend):
+    """DynamoDB Backend Adapter"""
+
+    def __init__(
+        self, mosaicid: str, region: str = os.getenv("AWS_REGION", "us-east-1")
+    ):
+        self.client = boto3.resource("dynamodb", region_name=region)
+        self.table = self.client.Table(mosaicid)
+        self.mosaic_def = self.fetch_mosaic_definition()
+
+    def tile(self, x: int, y: int, z: int, bucket: str, key: str) -> Tuple[str]:
+        """Retrieve assets for tile."""
+
+        return self.get_assets(x, y, z)
+
+    def point(self, lng: float, lat: float) -> Tuple[str]:
+        """Retrieve assets for point."""
+
+        min_zoom = self.mosaic_def["minzoom"]
+        quadkey_zoom = self.mosaic_def.get("quadkey_zoom", min_zoom)  # 0.0.2
+        tile = mercantile.tile(lng, lat, quadkey_zoom)
+        return self.get_assets(tile.x, tile.y, tile.z)
+
+    @functools.lru_cache(maxsize=512)
+    def fetch_mosaic_definition(self) -> Dict:
+        """Get Mosaic definition info."""
+        mosaic_def = fetch_dynamodb("-1")
+
+        # Numeric values are loaded from DynamoDB as Decimal types
+        # Convert maxzoom, minzoom, quadkey_zoom to float/int
+        for key in ["minzoom", "maxzoom", "quadkey_zoom"]:
+            if mosaic_def.get(key):
+                mosaic_def[key] = int(mosaic_def[key])
+
+        # Convert bounds, center to float/int
+        for key in ["bounds", "center"]:
+            if mosaic_def.get(key):
+                mosaic_def[key] = list(map(float, mosaic_def[key]))
+
+        return mosaic_def
+
+    def get_assets(x: int, y: int, z: int) -> Tuple[str]:
+        min_zoom = self.mosaic_def["minzoom"]
+        quadkey_zoom = self.mosaic_def.get("quadkey_zoom", min_zoom)  # 0.0.2
+
+        mercator_tile = mercantile.Tile(x=x, y=y, z=z)
+
+        # get parent
+        if mercator_tile.z > quadkey_zoom:
+            depth = mercator_tile.z - quadkey_zoom
+            for i in range(depth):
+                mercator_tile = mercantile.parent(mercator_tile)
+            quadkey = [mercantile.quadkey(*mercator_tile)]
+
+        # get child
+        elif mercator_tile.z < quadkey_zoom:
+            depth = quadkey_zoom - mercator_tile.z
+            mercator_tiles = [mercator_tile]
+            for i in range(depth):
+                mercator_tiles = sum(
+                    [mercantile.children(t) for t in mercator_tiles], []
+                )
+
+            mercator_tiles = list(filter(lambda t: t.z == quadkey_zoom, mercator_tiles))
+            quadkey = [mercantile.quadkey(*tile) for tile in mercator_tiles]
+        else:
+            quadkey = [mercantile.quadkey(*mercator_tile)]
```

Comment:
Can we make this block (get_parent, get_child...) an `.utils` function ?

Reply:
Yeah. I think that's good. I get a little confused by the itertools recursion. 

The function in `utils` can take an `accessor` argument that defaults to `.get()` for a `dict` but could be replaced by `.fetch_dynamodb()`

Reply:
yeah we don't need to over simplified!

Reply:
I refactored the quadkey finding into a separate function.